plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2021-2023 Jason Gerecke <jason.gerecke@wacom.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-wac-common.h"
+#include "fu-wac-device.h"
+#include "fu-wac-module-bluetooth-id9.h"
+#include "fu-wac-struct.h"
+
+struct _FuWacModuleBluetoothId9 {
+	FuWacModule parent_instance;
+};
+
+G_DEFINE_TYPE(FuWacModuleBluetoothId9, fu_wac_module_bluetooth_id9, FU_TYPE_WAC_MODULE)
+
+#define FU_WAC_MODULE_BLUETOOTH_ID9_PAYLOAD_SZ	   256
+#define FU_WAC_MODULE_BLUETOOTH_ID9_START_NORMAL   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL	   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE  0xFE
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_RAM	   0x02
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_BEGIN   0x03
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_DATA	   0x04
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL 0xEDB88320
+
+typedef struct __attribute__((packed)) {
+	guint16 unknown1;
+	guint32 unknown2;
+	guint16 size;
+} FuWacModuleBluetoothId9UnknownCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 spi_cmd;
+	guint32 start_addr;
+	guint16 size;
+	FuWacModuleBluetoothId9UnknownCmd data;
+
+} FuWacModuleBluetoothId9SpiCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 command;
+	guint16 size;
+	guint32 crc;
+	FuWacModuleBluetoothId9SpiCmd data;
+} FuWacModuleBluetoothId9LoaderCmd;
+
+static GBytes *
+fu_wac_module_bluetooth_id9_get_startcmd(const guint8 *data, gsize sz, gboolean full_erase)
+{
+	FuWacModuleBluetoothId9LoaderCmd cmd = {
+	    .command = full_erase ? FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE
+				  : FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL,
+	    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data) + sz)),
+	    .crc = 0,
+	    .data =
+		{
+		    .spi_cmd = 0x91,
+		    .start_addr = 0,
+		    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data.data) + sz)),
+		    .data =
+			{
+			    .unknown1 = GUINT16_TO_BE((guint16)0x7050),
+			    .unknown2 = 0,
+			    .size = GUINT16_TO_BE((guint16)sz),
+			},
+		},
+	};
+	guint32 crc;
+
+	// CRC(append(cmd.data, *data))
+	crc = fu_crc32_full((const guint8 *)&cmd.data,
+			    sizeof(cmd.data),
+			    ~0,
+			    FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	crc = fu_crc32_full(data, sz, ~crc, FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	cmd.crc = GUINT32_TO_BE(crc);
+
+	return g_bytes_new(&cmd, sizeof(cmd));
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_block(FuWacModule *self,
+					guint8 phase,
+					const guint8 *data,
+					gsize sz,
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GBytes) blob_chunk = NULL;
+	gsize buf_sz = sz + 1;
+	g_autofree guint8 *buf = g_malloc(buf_sz);
+
+	if (buf == NULL)
+		return FALSE;
+
+	buf[0] = phase;
+	memcpy(buf + 1, data, sz);
```

Comment:
I think g_autoptr(GByteArray) buf = g_byte_array_new(); fu_byte_array_append_uint8(buf, phase), g_byte_array_append(buf, data, sz) would be much nicer here

Reply:
Ack.